Feature/product cart

InJunKangW opened the pull request:
cart service 추가.

및 일부 서비스 메서드 수정 -  웹에서 api로 요청을 보낼 때 자신의 clientId를 보낼 수 있다는 가정 하에 짰던 코드들 수정 (웹 내부 자체에서는 clientId 확인 불가능하다고 합니다)

security config 조금 수정 - cart service 중 guest 관련 말곤 로그인 된 유저만 사용 가능하게.

refactor: 도서 등록/업데이트/조회 dto 수정, feat:카테고리 필터링 변경
1. 도서 등록/업데이트/조회 dto 필드의 카테고리명과 태그명의 타입을 List<String> -> Set<String>으로 변경.

2. 카테고리 필터링 변경
기존 필터링의 경우 딱 지정된 카테고리에 속해야만 결과에 반영
(상위 카테고리를 검색했을 때 하위 카테고리에 속해있는 도서 상품이 나오지 않음)

특정 카테고리를 골랐을 때 그 모든 하위 카테고리에 속한 도서 상품이 반환됨.
(그래서 필터링 조건으로 여러 카테고리를 동시에 두는 건 과부화 문제도 생길 것 같아서 제외)

InJunKangW committed:
feat: 장바구니 관련 레포지터리/서비스/컨트롤러 추가, 유틸 메서드 추가, 쿼리 개선

PageableUtil - pageNumCheck 메서드 추가. (요청한 페이지가 실제 페이지 보다 클 경우 예외 발생)

레포지터리의 @Query 어노테이션 최대한 제거.

InJunKangW committed:
feat: 회원/비회원 장바구니
회원은 DB에 저장
비회원은 웹 쿠키에 상품 아이디/수량만 저장 후 상품 아이디에 따른 상품 정보만 조회하게 함

gihwanJang commented on src/main/java/com/nhnacademy/bookstoreinjun/config/SecurityConfig.java:
```diff
@@ -29,6 +29,8 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http, Authentication
+                .addFilterBefore(new IdHeaderFilter("/api/product/client/**",DELETE), UsernamePasswordAuthenticationFilter.class)
                 .addFilterBefore(new RoleHeaderFilter("/api/product/admin/**",GET, ADMIN), UsernamePasswordAuthenticationFilter.class)
                 .addFilterBefore(new RoleHeaderFilter("/api/product/admin/**",POST, ADMIN), UsernamePasswordAuthenticationFilter.class)
                 .addFilterBefore(new RoleHeaderFilter("/api/product/admin/**",PUT, ADMIN), UsernamePasswordAuthenticationFilter.class)
```
최신화된 필터를 사용하시면 좋을 것 같습니다.
해당 필터는 다중권한을 지원하지 않습니다.

InJunKangW replied:
네 확인 후 수정하였습니다

InJunKangW committed:
refactor: 도서 등록/업데이트/조회 dto 수정, feat:카테고리 필터링 변경
1. 도서 등록/업데이트/조회 dto 필드의 카테고리명과 태그명의 타입을 List<String> -> Set<String>으로 변경.

2. 카테고리 필터링 변경
기존 필터링의 경우 딱 지정된 카테고리에 속해야만 결과에 반영
(상위 카테고리를 검색했을 때 하위 카테고리에 속해있는 도서 상품이 나오지 않음)

특정 카테고리를 골랐을 때 그 모든 하위 카테고리에 속한 도서 상품이 반환됨.
(그래서 필터링 조건으로 여러 카테고리를 동시에 두는 건 과부화 문제도 생길 것 같아서 제외)

InJunKangW committed:
feat: 웹에서 회원 장바구니 쿠키 날려먹었을 시 복구 해주는 서비스 추가
refactor: BookProductGetResponseDto, CartGetResponseDto 에
카테고리 아이디/이름, 태그 아이디/이름
을 갖는 Map 필드를 갖도록 변경

bakhuiwon326 commented on src/main/java/com/nhnacademy/bookstoreinjun/repository/BookQuerydslRepositoryImpl.java:
```diff
@@ -119,6 +129,57 @@ private Page<BookProductGetResponseDto> makePage(JPQLQuery<Tuple> query, JPQLQue
+    @Override
+    public Page<BookProductGetResponseDto> findAllBookPage(Pageable pageable, int productState){
+        OrderSpecifier<?> orderSpecifier = makeOrderSpecifier(pageable, "book");
+        BooleanBuilder whereBuilder = new BooleanBuilder();
```
boolean 대신 BooleanBulder 사용하신 이유가 궁금합니다
